src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -1282,4 +1282,89 @@ BOOST_AUTO_TEST_CASE(SimpleMeanSquaredErrorLayerTest)
   BOOST_REQUIRE_EQUAL(output.n_elem, 1);
 }
 
+/*
+ * Simple test for the BilinearInterpolation layer
+ */
+BOOST_AUTO_TEST_CASE(SimpleBilinearInterpolationLayerTest)
+{
+  arma::mat input, output, unzoomedOutput, expectedOutput;
+  size_t inRowSize = 10;
+  size_t inColSize = 10;
+  size_t outRowSize = 20;
+  size_t outColSize = 20;
+  size_t depth = 3;
+  input.zeros(inRowSize * inColSize *  depth, 1);
+  for (size_t d = 0; d < depth; d++)
+    for (size_t i = 0; i < inRowSize; i++)
+      for (size_t j = 0; j < inColSize; j++)
+      {
+        size_t ptr = d * inRowSize * inColSize + j * inColSize + i;
+        input[ptr] =  1 * (double)i / inRowSize +
+            1 * double(j) / inColSize + 1 * double(d) / depth;
+      }
+
+  expectedOutput.set_size(outRowSize * outColSize * depth, 1);
+  for (size_t d = 0; d < depth; d++)
+    for (size_t i = 0; i < outRowSize; i++)
+      for (size_t j = 0; j < outColSize; j++)
+      {
+        size_t ptr = d * outRowSize * outColSize + j * outColSize + i;
+        expectedOutput[ptr] = 1 * i / (double)outRowSize +
+            1 * j / (double)outColSize + 1 * d / (double)depth;
+      }
+
+  BilinearInterpolation<> layer(inRowSize, inColSize, outRowSize, outColSize,
+                                depth);
+
+  layer.Forward(std::move(input), std::move(output));
+  CheckMatrices(output - expectedOutput, arma::zeros(output.n_rows), 1e-12);
+
+
+  layer.Backward(std::move(output), std::move(output),
+        std::move(unzoomedOutput));
+  CheckMatrices(unzoomedOutput - input, arma::zeros(input.n_rows), 1e-12);
+}
+
+/**
+ * BilinearInterpolation layer numerical gradient test.
+ */
+BOOST_AUTO_TEST_CASE(GradientBilinearInterpolationLayerTest)
+{
+  // Add function gradient instantiation.
+  struct GradientFunction
+  {
+    GradientFunction()
+    {
+      input = arma::randu(10, 10);
+      target = arma::mat("1");
+
+      model = new FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization>(
+          input, target);
+      model->Add<IdentityLayer<> >();
+      model->Add<BilinearInterpolation<> >(10, 10, 20, 20, 1);
+      model->Add<LogSoftMax<> >();
+    }
+
+    ~GradientFunction()
+    {
+      delete model;
+    }
+
+    double Gradient(arma::mat& gradient) const
+    {
+      arma::mat output;
+      double error = model->Evaluate(model->Parameters(), 0, 1);
+      model->Gradient(model->Parameters(), 0, gradient, 1);
+      return error;
+    }
+
+    arma::mat& Parameters() { return model->Parameters(); }
+
+    FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization>* model;
+    arma::mat input, target;
+  } function;
+
+  BOOST_REQUIRE_LE(CheckGradient(function), 1e-4);
+}
```

Comment:
@zoq I'm not confident of this numerical test. Kris' original tests work fine, so probably the implementation is correct. A little help?

Reply:
Ping @zoq.

Reply:
Did you check with the original code? Perhaps we missed something? I guess, we have to go through each line and make sure it works as expected. Might be helpful to check the output of the gradient and backward method.

Reply:
@zoq I checked the core algorithm code a number of times, and found no problems, the implementation is most likely correct. The issue is with this test case, where for some reason only the first column of the input image is being passed to the `BilinearInterpolation` Layer (possibly because of the model chosen?). Can you help me in correcting this?